src/hotspot/share/gc/z/zUncommitter.cpp

```diff
@@ -94,3 +174,277 @@ void ZUncommitter::terminate() {
+bool ZUncommitter::uncommit_cycle_is_finished() const {
+  return _to_uncommit == 0;
+}
+
+bool ZUncommitter::uncommit_cycle_is_active() const {
+  return _cycle_start != 0.0;
+}
+
+bool ZUncommitter::uncommit_cycle_is_canceled() const {
+  return _cancel_time != 0.0;
+}
+
+size_t ZUncommitter::uncommit() {
+  precond(uncommit_cycle_is_active());
+
+  ZArray<ZVirtualMemory> flushed_vmems;
+  size_t flushed = 0;
+
+  {
+    // We need to join the suspendible thread set while manipulating capacity
+    // and used, to make sure GC safepoints will have a consistent view.
+    SuspendibleThreadSetJoiner sts_joiner;
+    ZLocker<ZLock> locker(&_partition->_page_allocator->_lock);
+
+    if (uncommit_cycle_is_canceled()) {
+      // We have committed within the delay, stop uncommitting.
+      return 0;
+    }
+
+    // We flush out and uncommit chunks at a time (~0.8% of the max capacity,
+    // but at least one granule and at most 256M), in case demand for memory
+    // increases while we are uncommitting.
+    const size_t current_max_capacity = _partition->_current_max_capacity;
+    const size_t limit_upper_bound = MAX2(ZGranuleSize, align_down(256 * M / ZNUMA::count(), ZGranuleSize));
+    const size_t limit = MIN2(align_up(current_max_capacity >> 7, ZGranuleSize), limit_upper_bound);
+
+    ZMappedCache& cache = _partition->_cache;
+
+    // Never uncommit more than the current uncommit watermark,
+    // (adjusted by what has already been uncommitted).
+    const size_t allowed_to_uncommit = MAX2(cache.min_size_watermark(), _uncommitted) - _uncommitted;
+    const size_t to_uncommit = MIN2(_to_uncommit, allowed_to_uncommit);
+
+    // Never uncommit below min capacity.
+    const size_t retain = MAX2(_partition->_used, _partition->_min_capacity);
+    const size_t release = _partition->_capacity - retain;
+    const size_t flush = MIN3(release, limit, to_uncommit);
+
+    // Flush memory from the mapped cache for uncommit
+    flushed = cache.remove_for_uncommit(flush, &flushed_vmems);
+    if (flushed == 0) {
+      // Nothing flushed
+      cancel_uncommit_cycle();
+      return 0;
+    }
+
+    // Record flushed memory as claimed and how much we've flushed for this partition
+    Atomic::add(&_partition->_claimed, flushed);
+  }
+
+  // Unmap and uncommit flushed memory
+  for (const ZVirtualMemory vmem : flushed_vmems) {
+    _partition->unmap_virtual(vmem);
+    _partition->uncommit_physical(vmem);
+    _partition->free_physical(vmem);
+    _partition->free_virtual(vmem);
+  }
+
+  {
+    SuspendibleThreadSetJoiner sts_joiner;
+    ZLocker<ZLock> locker(&_partition->_page_allocator->_lock);
+
+    // Adjust claimed and capacity to reflect the uncommit
+    Atomic::sub(&_partition->_claimed, flushed);
+    _partition->decrease_capacity(flushed, false /* set_max_capacity */);
+    register_uncommit(flushed);
+  }
+
+  return flushed;
+}
+
+void ZUncommitter::force_uncommit() {
+  ZLocker<ZConditionLock> locker(&_lock);
+  if (ZUncommit) {
```

Comment:
Since `ZUncommit` is constant would be better to check it before taking the lock